test/Hpack/ConfigSpec.hs

```diff
@@ -71,41 +82,26 @@ spec = do
-
-    it "adds the Paths_* module to the other-modules when no modules are specified" $ do
-      let exposed = ModuleSpecification Nothing Nothing
-          other = ModuleSpecification Nothing Nothing
-      determineModules ["Paths_foo"] [] exposed other `shouldBe` ([], ["Paths_foo"], [])
-
-    context "when the Paths_* module is part of the exposed-modules" $ do
-      it "does not add the Paths_* module to the other-modules" $ do
-        let exposed = ModuleSpecification ["Foo", "Paths_foo"] Nothing
-            other = ModuleSpecification Nothing Nothing
-        determineModules ["Paths_foo"] [] exposed other `shouldBe` (["Foo", "Paths_foo"], [], [])
-
-    it "includes all generated modules when exposed-modules and other-modules are Nothing" $ do
-      let exposed = ModuleSpecification Nothing ["ABC"]
-          other = ModuleSpecification Nothing ["XYZ"]
-      determineModules ["Paths_foo"] [] exposed other `shouldBe` (["ABC"], ["Paths_foo", "XYZ"], ["XYZ", "ABC"])
-
-    it "includes all generated modules when given explicit exposed-modules" $ do
-      let exposed = ModuleSpecification ["Foo"] ["ABC"]
-          other = ModuleSpecification Nothing ["XYZ"]
-      determineModules ["Paths_foo"] [] exposed other `shouldBe` (["Foo", "ABC"], ["Paths_foo", "XYZ"], ["XYZ", "ABC"])
-
-    it "includes all generated modules when given explicit exposed-modules and other-modules" $ do
-      let exposed = ModuleSpecification ["Foo"] ["ABC"]
-          other = ModuleSpecification ["Internal"] ["XYZ"]
-      determineModules ["Paths_foo"] [] exposed other `shouldBe` (["Foo", "ABC"], ["Internal", "XYZ"], ["XYZ", "ABC"])
-
-    it "includes all generated modules when given explicit other-modules" $ do
-      let exposed = ModuleSpecification Nothing ["ABC"]
-          other = ModuleSpecification ["Internal"] ["XYZ"]
-      determineModules ["Paths_foo"] [] exposed other `shouldBe` (["ABC"], ["Internal", "XYZ"], ["XYZ", "ABC"])
+    let
+      def :: DetermineModules
+      def = DetermineModules ["Paths_foo"] mempty mempty mempty mempty mempty
+
+    context "without inferable modules" $ do
+      withDepth 4 $ do
+        it "does not drop any modules that were explicitly specified (even if there are duplicates)" $ do
+          property $ \ exposedModules generatedExposedModules otherModules generatedOtherModules -> do
+            let
+              d = def {
+                _exposedModules = exposedModules
+              , _generatedExposedModules = generatedExposedModules
+              , _otherModules = otherModules
+              , _generatedOtherModules = generatedOtherModules
+              }
+            determineModules d `shouldBe` (
+                fromMaybeList exposedModules <> fromMaybeList generatedExposedModules
+              , maybe ["Paths_foo"] fromList otherModules <> fromMaybeList generatedOtherModules
+              , fromMaybeList (generatedExposedModules <> generatedOtherModules)
+              )
```

Comment:
In terms of code penetration, this test does everything the previous tests did.

We didn't test anything related to inferable modules though.

Reply:
My only worry here is that this is quite white-box, but I guess the end-to-end tests will exercise determineModules in a black-box way.